### DISCUSSION TITLE
Code Review: a lot of if else

### DISCUSSION THREAD

**Discussion by cd155** [JacquesCarette/Drasil#3289] [Show and tell]
Hi, I am currently taking an online class, and I would like to hear some feedback on the code I made. The code is made by Standard ML.

In the following three functions, each of them takes two arguments and returns a bool. Two arguments, named date1 and date2, are a tuple with 3 `int` type. The first `int` is year, the second `int` is month, and the last `int` is day. The following three functions will check if date1 is order than date2. Which one do you prefer?

```sml
fun is_older (date1: int*int*int, date2: int*int*int) = 
    if #1 date1 < #1 date2
    then true
    else if #1 date1 > #1 date2
    then false
    else (* when year equal *)
        if #2 date1 < #2 date2
        then true
        else if #2 date1 > #2 date2
        then false
        else (* when month equal *)
            if #3 date1 < #3 date2
            then true
            else if #2 date1 > #2 date2
            then false
            else false (* when day equal *)

fun is_older' (date1: int*int*int, date2: int*int*int) = 
    if #1 date1 = #1 date2
    then
        if #2 date1 = #2 date2
        then 
            if #3 date1 = #3 date2
            then
                false
            else
                #3 date1 < #3 date2
        else
            #2 date1 < #2 date2
    else 
        #1 date1 < #1 date2

fun is_older'' (date1: int*int*int, date2: int*int*int) = 
    let
        val year = (#1 date1, #1 date2)
        val month = (#2 date1, #2 date2)
        val date = (#3 date1, #3 date2)
        val xs = [year, month, date]

        fun is_older_Helper (xs: (int*int) list) =
            if null xs 
            then false
            else
                if #1 (hd xs) = #2 (hd xs) 
                then true andalso is_older_Helper (tl xs)
                else #1 (hd xs) < #2 (hd xs)
    in
        is_older_Helper xs
    end
```

**Comment by JacquesCarette** [comment]
To me, the best version would take in 2 records, each with 3 fields, called year, month and day. Taking a 3-tuple of ints is a very old style of coding.

Then I would write a boolean predicate, using 'and', and not use if-then-else at all.

**Reply by cd155** [reply]
I just wonder how to avoid the if-then-else completely.

Suppose we have two records; while we are comparing two records, we first need to compare the year, then the month, and the last day.

**Reply by JacquesCarette** [reply]
```
 (x.year < y.year) || 
((x.year == y.year) && 
    ( (x.month < y.month) || 
      ( (x.month == y.month) && (x.day < y.day)) ) )
```

**Reply by JacquesCarette** [reply]
Preferably with proper comments...

**Reply by cd155** [reply]
I see. With a combination of `or` and `and`, it would work, thanks.

By the way, tuples are syntactic sugar of record, so technically, the code already uses record.

**Reply by JacquesCarette** [reply]
While it is true that tuples are like records with accessors `fst` and `snd` (or `.1` and `.2`), most people don't see it that way. Positional access doesn't have the same affordability (a technical word from HCI) as more meaningful field names.

**Reply by cd155** [reply]
I see, thanks for mentioning it.
